Modules/_ctypes/_ctypes.c

```diff
@@ -3774,19 +3792,37 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
         return NULL;
     }
 #else
+	/* dlerror() always returns the latest error.
+	 *
+	 * Clear the previous value before calling dlsym(),
+	 * to ensure we can tell if our call resulted in an error.
+	 */
+    (void)dlerror();
     address = (PPROC)dlsym(handle, name);
```

Comment:
This isn't thread-safe, but
- I don't see a way to make it thread-safe, since `dlerror()` can be invoked by non-Python code.
- it should be OK for getting the *message* of an error.

As for the locale, I'd assume that `dlerror` returns ASCII. It certainly cannot support *arbitrary* locales (e.g. UTF-16 doesn't work in “zero-terminated `char *`”).

Reply:
`dlerror()` is thread-safe (as in, it uses a thread local), at least according to the man pages.

Reply:
On Linux, yes, but [POSIX says](https://pubs.opengroup.org/onlinepubs/9699919799.2013edition/functions/dlerror.html): 
> It is implementation-defined whether or not the dlerror() function is thread-safe. A thread-safe implementation shall return only errors that occur on the current thread.


I was wrong elsewhere: an UTF-16 locale is nonsense.
AFAICS, in glibc the message is meant to be translated, but might not be: https://repo.or.cz/glibc.git/blob/HEAD:/elf/dl-lookup.c#l795
So, `PyUnicode_DecodeLocale` & `PyErr_SetObject` would make sense.

Reply:
@ZeroIntensity  I can take care of dealing with locale since you're more informed in threads than me :)